### PR TITLE
Feature/24 tests

### PR DESCRIPTION
**Pull request by kotoko**
#24

### CODE REVIEW

**Comment by kotoko** [comment]
Now it's working. Test does not check all json response, but checks order, title and pagination.

File: api/build.gradle.kts [hunk 1]
```diff
@@ -53,10 +53,10 @@ sourceSets {
     create("integrationTest") {
         withConvention(KotlinSourceSet::class) {
             kotlin.srcDir("src/it/kotlin")
-            resources.srcDir("src/it/resources")
-            compileClasspath += sourceSets["main"].output + sourceSets["test"].output
-            runtimeClasspath += sourceSets["main"].output + sourceSets["test"].output
         }
+        resources.srcDir("src/it/resources")
+        compileClasspath += sourceSets["main"].output + sourceSets["test"].output + configurations["testRuntimeClasspath"]
+        runtimeClasspath += output + compileClasspath + sourceSets["test"].runtimeClasspath
```
Comment (michalkowol) [inline api/build.gradle.kts]: could we use 
```kt
compileClasspath += sourceSets["main"].output + sourceSets["test"].output	
            runtimeClasspath += sourceSets["main"].output + sourceSets["test"].output
```

File: api/build.gradle.kts [hunk 2]
```diff
@@ -149,12 +149,17 @@ dependencies {
...
 
     detektPlugins("io.gitlab.arturbosch.detekt:detekt-formatting:1.0.0-RC16")
+
+    testImplementation("org.springframework.cloud:spring-cloud-contract-wiremock")
+    testImplementation("org.springframework.boot:spring-boot-starter-test")
 }
 
 tasks.wrapper {
-    gradleVersion = "5.5.1"
+    gradleVersion = "5.6.1"
```
Comment (michalkowol) [inline api/build.gradle.kts]: Please update `wrapper`

run `./gradlew wrapper`

File: api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.concurrent.TimeUnit.SECONDS
+import wiremock.org.apache.http.HttpHeaders.CONTENT_TYPE
+
+class FetchFromAPITest : IntegrationTest() {
+
+    @Autowired
+    private lateinit var databaseSetup: DatabaseSetup
+
+    @Value("\${thehub.sync.domains}")
+    private lateinit var syncDomains: String
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: not used

File: api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        jobsResult
+            .andExpect(status().isOk)
+            .andExpect(jsonPath("$.totalPages").value(1))
+            .andExpect(jsonPath("$.number").value(1))
+            .andExpect(jsonPath("$.size").isNumber)
+            .andExpect(jsonPath("$.first").value(true))
+            .andExpect(jsonPath("$.last").value(true))
+            .andExpect(jsonPath("$.totalElements").value(2))
+            .andExpect(jsonPath("$.numberOfElements").value(2))
+            .andExpect(jsonPath("$.data[${chiefMarketingIndex}].jobId").value("chief-marketing-officer-cmo-6"))
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: replace `chiefMarketingIndex` with `0` and `chiefFinancialIndex` with `1`. `.andExpect(jsonPath("$.data[${chiefFinancialIndex}].jobId")` does not look natural

Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: use hamcrest matchers

```ks
import org.hamcrest.Matchers.containsString

// ...

.andExpect(jsonPath("$.data[0].jobId", containsString("marketing")))
```

File: api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    private fun addHtmlMapping(url: String, fileName: String) {
+        stubFor(get(url).willReturn(
+            aResponse()
+                .withStatus(200)
+                .withHeader("Content-Type", TEXT_HTML_VALUE)
+                .withBodyFile(fileName)
+        ))
+    }
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: Move to `addJsonMapping` and `addHtmlMapping` to `WireMockSetup` object

File: api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    fun ConditionFactory.untilRequest(resultProvider: () -> ResultActions, resultChecker: (ResultActions) -> Unit): ResultActions {
+        return this.until(resultProvider) { r ->
+            try {
+                resultChecker(r)
+                true
+            } catch (e: java.lang.AssertionError) {
+                false
+            }
+        }
+    }
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: Move to separate class - `extensions.kt`

File: api/src/it/resources/application-test.properties [hunk 7]
```diff
@@ -1,2 +1,5 @@
 spring.datasource.url=jdbc:postgresql://localhost:5432/codeloop_test
+logging.WireMock=debug
```
Comment (michalkowol) [inline api/src/it/resources/application-test.properties]: please set it to `warn`

File: api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+                .withHeader(CONTENT_TYPE, APPLICATION_JSON_UTF8_VALUE)
+                .withBodyFile(fileName)
+        ))
+    }
+
+    private fun addHtmlMapping(url: String, fileName: String) {
+        stubFor(get(url).willReturn(
+            aResponse()
+                .withStatus(200)
+                .withHeader("Content-Type", TEXT_HTML_VALUE)
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: Please use `CONTENT_TYPE`

File: api/src/it/kotlin/eu/codeloop/thehub/WireMockConfig.kt [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+            .notifier(Slf4jNotifier(true))
+    }
+
+}
+
+fun addJsonMapping(url: String, fileName: String) {
+    WireMock.stubFor(WireMock.get(url).willReturn(
+            WireMock.aResponse()
+                    .withStatus(200)
+                    .withHeader(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON_UTF8_VALUE)
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/WireMockConfig.kt]: Import static - use `CONTENT_TYPE` and `APPLICATION_JSON_UTF8_VALUE`

File: api/src/it/kotlin/eu/codeloop/thehub/WireMockConfig.kt [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+fun addHtmlMapping(url: String, fileName: String) {
+    WireMock.stubFor(WireMock.get(url).willReturn(
+            WireMock.aResponse()
+                    .withStatus(200)
+                    .withHeader(HttpHeaders.CONTENT_TYPE, MediaType.TEXT_HTML_VALUE)
+                    .withBodyFile(fileName)
+    ))
+}
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/WireMockConfig.kt]: Please `*Config` files are for `@Configuration` and only one class should be here.
Please move `addJsonMapping` and `addHtmlMapping` to `object WireMockUtils`

File: api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+        jobsResult
+            .andExpect(status().isOk)
+            .andExpect(jsonPath("$.totalPages").value(1))
+            .andExpect(jsonPath("$.number").value(1))
+            .andExpect(jsonPath("$.size").isNumber)
+            .andExpect(jsonPath("$.first").value(true))
+            .andExpect(jsonPath("$.last").value(true))
+            .andExpect(jsonPath("$.totalElements").value(2))
+            .andExpect(jsonPath("$.numberOfElements").value(2))
+            .andExpect(jsonPath("$.data[${0}].jobId", containsString("marketing")))
```
Comment (michalkowol) [inline api/src/it/kotlin/eu/codeloop/thehub/sync/FetchFromAPITest.kt]: Please just use `$.data[0].jobId`

File: api/build.gradle.kts (hunk 1 above)
Comment (kotoko) [reply api/build.gradle.kts]: I think it's working. Right?

**Pull request merged**